share/shwap/p2p/bitswap/blockstore_metrics.go

```diff
@@ -0,0 +1,212 @@
+package bitswap
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	bstore "github.com/ipfs/boxo/blockstore"
+	blocks "github.com/ipfs/go-block-format"
+	"github.com/ipfs/go-cid"
+	ipld "github.com/ipfs/go-ipld-format"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+)
+
+const (
+	notFoundKey = "not_found"
+	failedKey   = "failed"
+)
+
+var (
+	meter                   = otel.Meter("bitswap_blockstore")
+	_     bstore.Blockstore = (*BlockstoreWithMetrics)(nil)
+)
+
+// BlockstoreWithMetrics is a blockstore that collects metrics on blockstore operations.
+type BlockstoreWithMetrics struct {
+	bs      bstore.Blockstore
+	metrics *metrics
+}
+
+type metrics struct {
+	delete      metric.Int64Counter
+	has         metric.Int64Counter
+	get         metric.Int64Counter
+	getSize     metric.Int64Counter
+	put         metric.Int64Counter
+	putMany     metric.Int64Counter
+	allKeysChan metric.Int64Counter
```

Comment:
We dont need this one realistically. We dont implement it and never call it

Reply:
The blockstore metrics are design as general purpose wrapper. Metrics wrapper is also used for Light node blockstore, which implements those operations.